csp/baselib.py

```diff
@@ -759,32 +880,101 @@ def comp(x: ts["T"], y: ts["U"]):
     return comp
 
 
-# Math operators
+@lru_cache(maxsize=512)
+def define_unary_op(name, op_lambda):
+    float_out_type, int_out_type, generic_out_type = [None] * 3
+    if name in COMP_OPS:
+        float_out_type = bool
+        int_out_type = bool
+        generic_out_type = bool
+    elif name in MATH_OPS:
+        float_out_type = float
+        if name in ("abs",):
+            int_out_type = int
+            generic_out_type = "T"
+        else:
+            int_out_type = float
+            generic_out_type = float
 
-add = define_op("add", lambda x, y: x + y)
-sub = define_op("sub", lambda x, y: x - y)
-multiply = define_op("multiply", lambda x, y: x * y)
-pow = define_op("pow", lambda x, y: x**y)
-divide = define_op("divide", lambda x, y: x / y)
-min = define_op("min", lambda x, y: x if x < y else y)
-max = define_op("max", lambda x, y: x if x > y else y)
+    from csp.impl.wiring.node import _node_internal_use
 
-# Comparison operators
+    @_node_internal_use(cppimpl=MATH_COMP_OPS_CPP.get((name, "float"), None), name=name)
+    def float_type(x: ts[float]) -> ts[float_out_type]:
+        if csp.valid(x):
+            return op_lambda(x)
 
-eq = define_op("eq", lambda x, y: x == y)
-ne = define_op("ne", lambda x, y: x != y)
-gt = define_op("gt", lambda x, y: x > y)
-lt = define_op("lt", lambda x, y: x < y)
-ge = define_op("ge", lambda x, y: x >= y)
-le = define_op("le", lambda x, y: x <= y)
+    @_node_internal_use(cppimpl=MATH_COMP_OPS_CPP.get((name, "int"), None), name=name)
+    def int_type(x: ts[int]) -> ts[int_out_type]:
+        if csp.valid(x):
+            return op_lambda(x)
 
-# Other math ops
+    numpy_func = MATH_COMP_OPS_CPP.get((name, "np"), op_lambda)
+
+    @_node_internal_use(name=name)
+    def numpy_type(x: ts["T"]) -> ts[np.ndarray]:
```

Comment:
You can do ts[np.ndarray] for the input type here, it should pick up both NDArrays and 1DArrays

Reply:
is one preferred over the other? I was just tweaking what was there

Reply:
I think it's generally better to be more explicit with regards to type checking. In this case it doesn't really matter since we're checking the type to be a Numpy array before returning the `numpy_type` impl node anyways. Still, I'd make the change for style reasons.